src/live_restore/live_restore_state.c

```diff
@@ -255,21 +255,77 @@ __wt_live_restore_get_state_string(WT_SESSION_IMPL *session, WT_ITEM *lr_state_s
     return (0);
 }
 
+/*
+ * __wt_live_restore_turtle_update --
+ *     Intercept updates to the turtle file so we can take the state lock first. The state lock must
+ *     be held for the entire process and taken before we take the turtle lock.
+ */
+int
+__wt_live_restore_turtle_update(
+  WT_SESSION_IMPL *session, const char *key, const char *value, bool take_turtle_lock)
+{
+    WT_DECL_RET;
+
+    WTI_LIVE_RESTORE_FS *lr_fs = (WTI_LIVE_RESTORE_FS *)S2C(session)->file_system;
+
+    bool reentrant = __wt_spin_owned(session, &lr_fs->state_lock);
+    if (!reentrant)
+        __wt_spin_lock(session, &lr_fs->state_lock);
+
+    if (take_turtle_lock)
+        WT_WITH_TURTLE_LOCK(session, ret = __wt_turtle_update(session, key, value));
+    else
+        ret = __wt_turtle_update(session, key, value);
+    WT_ERR(ret);
+
+err:
+    if (!reentrant)
+        __wt_spin_unlock(session, &lr_fs->state_lock);
+
+    return (ret);
+}
+
+/*
+ * __wt_live_restore_turtle_rewrite --
+ *     Intercept calls to rewrite the turtle file so we can take the state lock first. The state
+ *     lock must be held for the entire process and taken before we take the turtle lock.
+ */
+int
+__wt_live_restore_turtle_rewrite(WT_SESSION_IMPL *session)
+{
+    WT_DECL_RET;
+    WTI_LIVE_RESTORE_FS *lr_fs = (WTI_LIVE_RESTORE_FS *)S2C(session)->file_system;
+
+    bool reentrant = __wt_spin_owned(session, &lr_fs->state_lock);
+    if (!reentrant)
+        __wt_spin_lock(session, &lr_fs->state_lock);
+
+    WT_WITH_TURTLE_LOCK(session, ret = __wt_metadata_turtle_rewrite(session));
+    WT_ERR(ret);
+
+err:
+    if (!reentrant)
+        __wt_spin_unlock(session, &lr_fs->state_lock);
+
+    return (ret);
+}
+
 /*
  * __wti_live_restore_get_state --
  *     Get the live restore state. Take the state lock if it isn't already held.
  */
 WTI_LIVE_RESTORE_STATE
 __wti_live_restore_get_state(WT_SESSION_IMPL *session, WTI_LIVE_RESTORE_FS *lr_fs)
 {
```

Comment:
There's no functional changes to `__wti_live_restore_get_state`. I've just reworked it to use the same `bool reentrant` logic as the other functions for consistency.